Guides 2.0 - Actions # Control Flow section (#119)
hanami/guides: content/v2.0/actions/control-flow.md

---
title: Control Flow
order: 110
aliases:
  - "/actions/control-flow"
---

## Callbacks

If we want to execute some logic before and/or after `#handle` is executed, we can use a callback.

> Callbacks are useful to declutter code for common tasks like checking if a user is signed in, set a record, handle 404 responses or tidy up the response.

Reviewer: What do you mean by `set a record`?

Author: I see it like: fetching a record to use in order to process/check authorize access to. This is not changed from guides 1.3.

Author: I have removed this mention for clarity. It's not necessary to show the point.

The corresponding DSL methods are `before` and `after`.

> These methods each accept a symbol that is the name of the method that we want to call, or an anonymous proc.

Reviewer: ```suggestion
These methods each accept a symbol that is the name of the method that we want to call, or a proc.
```

### Methods

```ruby
# apps/actions/books/show.rb

module Bookshelf
  module Actions
    module Books
      class Index < Action
        before :validate_params
        
        params do
          required(:id).filled(:integer)
        end

        private

        def validate_params(request, response)
          params = request.params
          halt 422, params.errors.to_h unless params.valid?
        end
      end
    end
  end
end
```

With the code above, we are validating and coercing URL `:id` parameter, making sure it will always be integer. Otherwise, we stop request processing, immediately returning error to the browser.

Because it isn't strictly related to our business logic, we move it to a callback.

### Proc

The example above can be rewritten with anonymous proc.

They are bound to the instance context of the action.

```ruby
# app/actions/books/show.rb
module Bookshelf
  module Actions
    module Books
      class Show < Action

>         before { |req, res| halt 422, req.params.errors.to_h unless req.params.valid? }

Reviewer: ```suggestion
        before { |request, response| halt 422, request.params.errors.to_h unless request.params.valid? }
```

          halt 404, "These aren't the droids you're looking for"
        end
      end
    end
  end
end
```

When `#halt` is used, **Hanami** renders a default status page with the HTTP status and the message.

<p><img src="/v2.0/actions/404-response.png" alt="Hanami 404 response" class="img-responsive"></p>

## Redirect

A special case of control flow management is relative to HTTP redirect.
If we want to reroute a request to another resource we can use `redirect_to` called on the response object.

When `redirect_to` is invoked, control flow is stopped and **subsequent code in the action is not executed**.

It accepts a string that represents an URI, and an optional `:status` argument.
By default the status is set to `302`.

```ruby
# app/actions/books/index.rb

module Bookshelf
  module Actions
    module Books
      class Index < Action
        def handle(request, response)
          response.redirect_to routes.path(:root), status: 302
          raise "This line will never be executed"
        end
      end
    end
  end
end
```

### Back

Sometimes you'll want to `redirect_to` back in your browser's history so the easy way to do it
is the following way:

```ruby

> res.redirect_to request.get_header("Referer") || fallback_url

Reviewer: ```suggestion
response.redirect_to request.get_header("Referer") || fallback_url
```

```ruby
# app/actions/books/index.rb

module Bookshelf
  module Actions
    module Books
      class Index < Action
        before :authenticate_user!

        def handle(request, response)
          # ...
        end

        private

        def authenticate_user!(request, response)
          halt 401 unless request.session[:user_id]
        end
      end
    end
  end
end
```

`#halt` accepts an HTTP status code as the first argument and an optional body as its second argument. If no body is provided, the body will be set to a message corresponding to the status code (for example the body of a `401` response will be `"Unauthorized"`).

```ruby
# app/actions/books/index.rb

module Bookshelf
  module Actions
    module Books
      class Index < Action
        def handle(request, response)
          halt 404, "These aren't the droids you're looking for"
        end
      end
    end
  end
end
```

<p><img src="/v2.0/actions/404-response.png" alt="Hanami 404 response" class="img-responsive"></p>

## Redirects

> To redirect a request to another location, call the `#redirect_to` method on the `request` object.

Reviewer 2: ```suggestion
To redirect a request to another location, call the `#redirect_to` method on the `response` object.
```

> When `redirect_to` is invoked, control flow is stopped and subsequent code in the action is not executed.

Reviewer 2: ```suggestion
When you call `redirect_to`, control flow is stopped and subsequent code in the action is not executed.
```
